### ISSUE TITLE
asn1map usage

### ISSUE THREAD

**Issue by shauny1** [phpseclib/phpseclib#1211]
I'm having a bit of trouble getting asn1map to work and have not had any success in finding any documentation on it.

I want to parse CMS CompressedData  content (following is the ASN1 information from various RFCs):

`
CMSVersion ::= INTEGER
                     { v0(0), v1(1), v2(2), v3(3), v4(4), v5(5) }
	
	EncapsulatedContentInfo ::= SEQUENCE {
        eContentType ContentType,
        eContent [0] EXPLICIT OCTET STRING OPTIONAL }

      ContentType ::= OBJECT IDENTIFIER
	
	CompressedData ::= SEQUENCE {
        version CMSVersion,
        compressionAlgorithm CompressionAlgorithmIdentifier,
        encapContentInfo EncapsulatedContentInfo
        }
		
	CompressionAlgorithmIdentifier ::= AlgorithmIdentifier

    Algorithm Identifiers

   id-alg-zlibCompress OBJECT IDENTIFIER ::= { iso(1) member-body(2)
       us(840) rsadsi(113549) pkcs(1) pkcs-9(9) smime(16) alg(3) 8 }

    Content Type Object Identifiers

   id-ct-compressedData OBJECT IDENTIFIER ::= { iso(1) member-body(2)
       us(840) rsadsi(113549) pkcs(1) pkcs-9(9) smime(16) ct(1) 9 }

		
	EncapsulatedContentInfo ::= SEQUENCE {
        eContentType ContentType,
        eContent [0] EXPLICIT OCTET STRING OPTIONAL }

      ContentType ::= OBJECT IDENTIFIER
`

I tried (and failed) to use the following mapping:

`$asnEncapsulatedContentInfo = array('type' => FILE_ASN1_TYPE_SEQUENCE, 'children' => array('eContentType' => array('type' => FILE_ASN1_TYPE_OBJECT_IDENTIFIER), 'eContent' => array('type' => FILE_ASN1_TYPE_OCTET_STRING)));

$asnCompressionAlgorithm = array('type' => FILE_ASN1_TYPE_SEQUENCE, 'children' => array('algorithm' => array('type' => FILE_ASN1_TYPE_OBJECT_IDENTIFIER), 'parameters' => array('type' => FILE_ASN1_TYPE_OBJECT_IDENTIFIER)));

$asnCMSVersion = array('type' => ASN1_File_Type_Integer);

$asnCompressedData = array('type' => FILE_ASN1_TYPE_SEQUENCE, 'children' => array('version' => array('type' => FILE_ASN1_TYPE_INTEGER), 'compressionAlgorithm' => $asnCompressionAlgorithm, 'encapContentInfo' =>  asnEncapsulatedContentInfo));`

I tried to use $asnCompressedData as my $mapping for asn1map. decodeBER() returns a $decoded array but asn1map returns null using the above mapping. Obviously some error with my mapping but I'm also unsure of the purpose of asn1map too.

If I'm only interested in getting the equivalent of the HEXDUMP content from openssl asn1parse command for the OCTET STRING is the above approach the best way using phplibsect?  If so how do I construct my $mapping? And if not what other way is there using this library?

**Comment by terrafrost** [comment]
One thing that might be beneficial for you to do is to build off of https://github.com/terrafrost/phpseclib/tree/cms as that already does some CMS stuff.

Anyway, 

**asnEncapsulatedContentInfo**

> ```
> EncapsulatedContentInfo ::= SEQUENCE {
>     eContentType ContentType,
>     eContent [0] EXPLICIT OCTET STRING OPTIONAL }
> 
>   ContentType ::= OBJECT IDENTIFIER
> ```

> $asnEncapsulatedContentInfo = array('type' => FILE_ASN1_TYPE_SEQUENCE, 'children' => array('eContentType' => array('type' => FILE_ASN1_TYPE_OBJECT_IDENTIFIER), 'eContent' => array('type' => FILE_ASN1_TYPE_OCTET_STRING)));

In the actual ASN1 eContent is optional so you'd need to do this:

```php
$asnEncapsulatedContentInfo = array(
        'type'     => FILE_ASN1_TYPE_SEQUENCE,
        'children' => array(
            'eContentType'     => array('type' => FILE_ASN1_TYPE_OBJECT_IDENTIFIER),
            'eContent' => array(
                'type' => FILE_ASN1_TYPE_OCTET_STRING,
                'constant' => 0,
                'optional' => true,
                'explicit' => true
            )
        )
);
```

**asnCMSVersion**

```php
$asnCMSVersion = array('type' => ASN1_File_Type_Integer);
```
That should be all caps. And you got ASN1 and File switched. Should be FILE_ASN1_TYPE_INTEGER.

**asnCompressionAlgorithm**

AlgorithmIdentifier is pretty much always the same. Here's how I do it in File/X509.php and File/CMS.php:

```php

        // http://tools.ietf.org/html/rfc5280#section-4.1.1.2
        $AlgorithmIdentifier = array(
            'type'     => FILE_ASN1_TYPE_SEQUENCE,
            'children' => array(
                'algorithm'  => array('type' => FILE_ASN1_TYPE_OBJECT_IDENTIFIER),
                'parameters' => array(
                                    'type'     => FILE_ASN1_TYPE_ANY,
                                    'optional' => true
                                )
            )
        );
```

Note that the master branch greatly facilitates re-use. In https://github.com/terrafrost/phpseclib/tree/cms $AlgorithmIdentifier is defined twice - once in CMS.php and once in X509.php. In the master branch auto-loading is utilized such that AlgorithmIdentifier only ever needs to be defined once:

https://github.com/phpseclib/phpseclib/blob/master/phpseclib/File/ASN1/Maps/AlgorithmIdentifier.php

**asnCompressedData**

```php
$asnCompressedData = array('type' => FILE_ASN1_TYPE_SEQUENCE, 'children' => array('version' => array('type' => FILE_ASN1_TYPE_INTEGER), 'compressionAlgorithm' => $asnCompressionAlgorithm, 'encapContentInfo' => asnEncapsulatedContentInfo));`
```

`asnEncapsulatedContentInfo` should have a dollar sign before it. eg. `$asnEncapsulatedContentInfo`

>  Obviously some error with my mapping but I'm also unsure of the purpose of asn1map too.

If you just decode the BER then all you know is that there's an integer but you don't know what that integer means. What does a particular octet string or bit string represent? There's not an easy way to know. `asn1map` basically maps out the binary data into a more useful / easily parsed data structure.

> If I'm only interested in getting the equivalent of the HEXDUMP content from openssl asn1parse command for the OCTET STRING is the above approach the best way using phplibsect?

If that's all you're doing then you could do something similar to http://phpseclib.sourceforge.net/x509/asn1parse.php . That utilizes phpseclib to produce output pretty much identical to OpenSSL's asn1parse. The code isn't super elegant but it suffices as a proof of concept.

**Comment by shauny1** [comment]
I have tried with your suggestions but not joy. My content is the base64 p7z part of a decrypted (but not uncompressed) SMIME/CMS message. Apologies for the missing $ and constant typos they were a transcribing mistake. The code now looks as follows:

        $ber = file_get_contents('cmsCompressed.base64');
        $asn1 = new File_ASN1();
        $decoded = $asn1->decodeBER(base64_decode($ber));
        $asnEncapsulatedContentInfo = array(
                'type' => FILE_ASN1_TYPE_SEQUENCE,
                'children' => array(
                        'eContentType' => array(
                                'type' => FILE_ASN1_TYPE_OBJECT_IDENTIFIER
                        ),
                        'eContent' => array(
                                'type' => FILE_ASN1_TYPE_OCTET_STRING,
                                'constant' => 0,
                                'optional' => true,
                                'explicit' => true
                        )
                )
        );

        $asnCompressionAlgorithm = array(
                'type' => FILE_ASN1_TYPE_SEQUENCE,
                'children' => array(
                        'algorithm' => array(
                                'type' => FILE_ASN1_TYPE_OBJECT_IDENTIFIER
                        ),
                        'parameters' => array(
                                'type' => FILE_ASN1_TYPE_ANY,
                                'optional' => true,
                        )
                )
        );

        $asnCMSVersion = array('type' => FILE_ASN1_TYPE_INTEGER);

        $asnCompressedData = array(
                'type' => FILE_ASN1_TYPE_SEQUENCE,
                'children' => array(
                        'version' => array(
                                'type' => FILE_ASN1_TYPE_INTEGER
                        ),
                        'compressionAlgorithm' => $asnCompressionAlgorithm,
                        'encapContentInfo' => $asnEncapsulatedContentInfo
                )
        );

        print_r($asnCompressedData);
        $mappedDetails = $asn1->asn1map($decoded[0], $asnCompressedData);

Now at this stage $mappedDetails is null. And this is what my content looks like with openssl:

 openssl asn1parse -in data.base64 -i

    0:d=0  hl=4 l= 814 cons: SEQUENCE
    4:d=1  hl=2 l=  11 prim:  OBJECT            :id-smime-ct-compressedData
   17:d=1  hl=4 l= 797 cons:  cont [ 0 ]
   21:d=2  hl=4 l= 793 cons:   SEQUENCE
   25:d=3  hl=2 l=   1 prim:    INTEGER           :00
   28:d=3  hl=2 l=  13 cons:    SEQUENCE
   30:d=4  hl=2 l=  11 prim:     OBJECT            :zlib compression
   43:d=3  hl=4 l= 771 cons:    SEQUENCE
   47:d=4  hl=2 l=   9 prim:     OBJECT            :pkcs7-data
   58:d=4  hl=4 l= 756 cons:     cont [ 0 ]
   62:d=5  hl=4 l= 752 prim:      OCTET STRING      [HEX DUMP]:DEADBEEFADADADADADA..

I can access the content I want with:

        $zlibContent = $decoded[0]['content'][1]['content'][0]['content'][2]['content'][1]['content'][0]['content'];

and then uncompress with:

gzuncompress($zlibContent);

This works but is, uhm, a bit hacky and not very robust?

Does the CMS implementation in phplibsec thus far offer the equivalent of openssl uncompress? Current versions of the openssl tool in Linux are no longer compiled against ZLIB and so I'm trying to mimic this command in PHP.

I will also check out the asn1parse.php you provided as a link.  Is this included in the phplibsec packages that are available in distros? If not I can just clone from here anyway.

If I find any issues playing with any of this stuff I'll be sure to report them here.

Great work by the way, I hope to get the asn1parsing working, I feel I'm close!

**Comment by terrafrost** [comment]
To provide further support what'd really help is the `cmsCompressed.base64` you're performing your tests on. With that I could perform my own tests.

> Does the CMS implementation in phplibsec thus far offer the equivalent of openssl uncompress?

No. Currently CMS.php in that branch only supports signing (with [CAdES](https://en.wikipedia.org/wiki/CAdES_(computing)) support). I mean, it could be extended to support it (utilizing gzuncompress like you did) but it doesn't currently.

One thing to keep in mind... that file isn't officially supported. It'll probably be implemented in master / 3.0 at some point but not any time soon.

> I will also check out the asn1parse.php you provided as a link. Is this included in the phplibsec packages that are available in distros?

It's in the phpseclib/docs repo:

https://github.com/phpseclib/docs/blob/master/x509/asn1parse.php

**Comment by shauny1** [comment]
[compressed.zip](https://github.com/phpseclib/phpseclib/files/1475531/compressed.zip)

**Comment by terrafrost** [comment]
Thanks! I probably won't be able to look until the weekend but I'll take a look!

**Comment by terrafrost** [comment]
The latest commit in https://github.com/terrafrost/phpseclib/tree/cms adds support for this:

https://github.com/terrafrost/phpseclib/commit/1f7b23c81ebdcd2c422e06cb7928efc8cb9e24a4#diff-cb8cd7734a8b95f8fc94ba2b7eae5da8

Sample code:

```
<?php
include('File/CMS.php');

$cms = new File_CMS();
$r = $cms->load('MIIDLgYLKoZIhvcNAQkQAQmgggMdMIIDGQIBADANBgsqhkiG9w0BCRADCDCCAwMG
CSqGSIb3DQEHAaCCAvQEggLweJydk1tvqzgUhd+R+A8orxUN4UAurfpgE4eQBFoI
kMBL5YC5NNwChpD8+kMqjeaiM6Ojsfy019pb39qWdU1HvEvqJi2LF27yLLCMUhaU
FJS3bxV54fI2o2mFazpu0rgg4StX1SUtgzJ7G+GqytIA06F3XJ2DZsY/PJi2NRm9
cvkgZfHbqEkwL8rToXIq2yLE9e1txA9nKsOpJM0FCU5XcyTIMwEKE7RYLqEiC2Ax
H7EMy/D871j/CU1JT8dVhtPiT2WZNlXZpPQ7KKAUB0k+1F+5KM1IgXPyRklDP5s8
zcknHcZ8BmVe1aRpSPhMwvRBYydpww0Xcw8vlw8ijsn/5/yPBb5y30yjb57nataM
/tJc46KJSM2jIijDtIhfuBNuyFT6l7T4V2n/NplldE2D1PQ229LXki4wgBko5VCr
vhQFYE0BJuqXd2DAOIuTcwx9U0cgRiuwBDsYny8sk5xTdXEVILyiPo5Xi8dEYmoQ
2HCvI6dX7mADY8OFwLPBeeXqln5Fprd0TVNbgspmmZM4yXauIXhHK9P36Lq+fqvb
JUw2p8LKglzOQgUe8cEq8NG6aysokD3UQ9VUFHM5rLfTXDgpoUt/TWo8SP/gHBKt
ANgPyQBqzeD9SbKHR26bZEHXp/H9YF02nbO+m7KWeG0igNIJNtq1Dx2vr+B853pp
eZTsbnu/AlmWFQhjk2QRy4hK5Vih9aTOvuotbhfuRRIuRJypXbTZHsUfZqmGSZCs
GzUFoCg3w8bmY1qEU9sZi7Gos0z/ETSXCdrSnRgpdh/PLsjzJf+rnbTd0a/WG1dQ
86eDY3/0+uJUT0DXGgpWZu+Gv0NadGIZEx5iIItYJOltddraoft1RCeK2sVc7zNj
lQn7KjqM89pWpL4kiGwiesGdqXt4HuzvCcvshWl9v4l12xqT8OCQdO5M5x9y5Ee4
3QbXD8l/t3yoqxFeF+BWyESwQAQIfOrQj+y8oCxzjkU3Q++gBLH6+1+E5x/en0Jj
cT0=');

print_r($r);
```

**Comment by terrafrost** [comment]
#2125 (targeting the yet to be released phpseclib 4.0) introduces built in support for this:

https://terrafrost.com/phpseclib/v4/docs/cms/compressed
